llama.cpp

```diff
@@ -17016,6 +17016,20 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>";
         }
+    } else if (tmpl == "llama3" || (tmpl.find("<|start_header_id|>") != std::string::npos && tmpl.find("<|end_header_id|>") != std::string::npos)) {
+        // Llama 3
+        bool first_message = true; // add BOS at the beginning
+        for (auto message : chat) {
+            std::string role(message->role);
+            if(first_message){
+                ss << "<|begin_of_text|>";
+                first_message = false;
+            }
+            ss << "<|start_header_id|>" << message->role << "<|end_header_id|>\n\n" << trim(message->content) << "<|eot_id|>";
```

Comment:
Or more simply
```suggestion
            ss << "<|start_header_id|>" << role << "<|end_header_id|>\n\n" << trim(message->content) << "<|eot_id|>";
```

Reply:
Yeah sorry about that, I'm literally not a C++ coder, did this just based on the other examples I could see.
For the most part (day job) I code in C#
Thanks for the feedback!

Reply:
No problem 😉